code/go/0chain.net/blobbercore/reference/ref.go

```diff
@@ -324,55 +426,55 @@ func (fr *Ref) GetFileHashData() string {
 }
 
 func (fr *Ref) CalculateFileHash(ctx context.Context, saveToDB bool) (string, error) {
+
 	fr.Hash = encryption.Hash(fr.GetFileHashData())
-	fmt.Println("file hash", fr.Path, fr.Hash)
 	fr.NumBlocks = int64(math.Ceil(float64(fr.Size*1.0) / float64(fr.ChunkSize)))
 	fr.PathHash = GetReferenceLookup(fr.AllocationID, fr.Path)
 	fr.PathLevel = len(GetSubDirsFromPath(fr.Path)) + 1
 	fr.LookupHash = GetReferenceLookup(fr.AllocationID, fr.Path)
+
 	var err error
 	if saveToDB {
-		err = fr.Save(ctx)
+		//err = fr.Save(ctx)
+		err = fr.SaveFile(ctx)
 	}
 	return fr.Hash, err
 }
 
 func (r *Ref) CalculateDirHash(ctx context.Context, saveToDB bool) (string, error) {
 	// empty directory, return hash directly
+
 	if len(r.Children) == 0 && !r.childrenLoaded {
 		return r.Hash, nil
 	}
 	sort.SliceStable(r.Children, func(i, j int) bool {
 		return strings.Compare(r.Children[i].LookupHash, r.Children[j].LookupHash) == -1
 	})
-	for _, childRef := range r.Children {
-		_, err := childRef.CalculateHash(ctx, saveToDB)
-		if err != nil {
-			return "", err
-		}
-	}
 	childHashes := make([]string, len(r.Children))
 	childPathHashes := make([]string, len(r.Children))
 	var refNumBlocks int64
 	var size int64
 	for index, childRef := range r.Children {
+		_, err := childRef.CalculateHash(ctx, saveToDB)
```

Comment:
@abouroumine I think `childRef.CalculateHash` is required only for affected refs. For unaffected refs we can simply get has value from `Hash` field.

We should avoid hash calculation where we can because it is computationally resource consuming.
It would be great to distinguish affected and unaffected rows. For example if path is `/a/b/c/d/e.txt` then hash calculation is basically required for refs with paths `/`, `/a`, `/a/b`, `/a/b/c`, `/a/b/c/d` and `/a/b/c/d/e.txt`. Correct me If I am wrong here. @abouroumine @moldis

Reply:
Should we then introduce a new parameter that identifies the Affected and Unaffected Refs ? Especially if we are dealing with a file update / copy / move.